{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2dade689",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7aa25811",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Streptomyces antibioticus\n",
      "Streptomyces aureofaciens\n",
      "Streptomyces avermitilis\n",
      "Streptomyces azureus\n",
      "Streptomyces bicolor\n",
      "Streptomyces cattleya\n",
      "Streptomyces coelicolor\n",
      "Streptomyces flavogriseus\n",
      "Streptomyces flavovirens\n",
      "Streptomyces griseofuscus\n",
      "Streptomyces griseus\n",
      "Streptomyces himastatinicus\n",
      "Streptomyces indigocolor\n",
      "Streptomyces lividans\n",
      "Streptomyces mirabilis\n",
      "Streptomyces platensis\n",
      "Streptomyces roseosporus\n",
      "Streptomyces sanglieri\n",
      "Streptomyces scabiei\n",
      "Streptomyces sp.\n",
      "Streptomyces toxytricini\n",
      "Streptomyces tricolor\n",
      "Streptomyces venezuelae\n",
      "Streptomyces virginiae\n",
      "Streptomyces viridochromogenes\n",
      "Streptomyces xanthochromogenes\n"
     ]
    }
   ],
   "source": [
    "#STEP 1 \n",
    "#GENERATE A LIST OF ALL STREPTOMYCES SPECIES\n",
    "\n",
    "response = requests.get(\"https://phagesdb.org/api/host_species/\")\n",
    "\n",
    "all_species = []\n",
    "for x in response.json():\n",
    "    if x[\"genus\"] == 'Streptomyces':\n",
    "        all_species.append(x[\"id\"])\n",
    "        print(x[\"genus\"] + \" \"+ x[\"species_name\"] )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "24fd519a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#STEP 2: Now that we found all the host, can we find the siphoviridae phages infecting these host\n",
    "\n",
    "all_phages = []\n",
    "for x in all_species:\n",
    "    response = requests.get(\"https://phagesdb.org/api/host_species/\" + str(x)+ \"/phagelist/\")\n",
    "    for x in response.json()[\"results\"]:\n",
    "        if (x['morphotype']) == 'SIPHO':\n",
    "            all_phages.append(x)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "626cb718",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "103\n"
     ]
    }
   ],
   "source": [
    "len(all_phages)\n",
    "c = 0\n",
    "valid_phages = []\n",
    "for x in all_phages:\n",
    "    if (x[\"em_file\"]) != None and x[\"genome_length\"] != None:\n",
    "        valid_phages.append(x)\n",
    "        c = c+1\n",
    "print(c) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bbff784a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'phage_name': 'Rowa',\n",
       " 'old_names': '',\n",
       " 'program': {'short_name': 'SEAPHAGES',\n",
       "  'program_name': 'Science Education Alliance-Phage Hunters Advancing Genomics and Evolutionary Science',\n",
       "  'program_blurb': '',\n",
       "  'program_URL': 'http://www.hhmi.org/grants/sea/index.html'},\n",
       " 'naming_notes': \"It's  a girl name in Arabic. It gives the meaning of saturation.I like this name and it might be the name of my daughter in future\",\n",
       " 'finder_name': 'Ahmad Sulaiman',\n",
       " 'found_city': 'New Braunfels',\n",
       " 'found_state': 'TX',\n",
       " 'found_country': 'USA',\n",
       " 'found_GPS_lat': '29.711028 N',\n",
       " 'found_GPS_long': '98.133833 W',\n",
       " 'found_year': 2016,\n",
       " 'p_institution': {'institution_code': 'UNTD',\n",
       "  'institution_name': 'University of North Texas',\n",
       "  'institution_logo': 'https://phagesdb.org/media/institutions/UNT-01.jpg',\n",
       "  'institution_url': 'http://www.unt.edu/'},\n",
       " 'enriched_sample': True,\n",
       " 'discovery_notes': 'The soil sample was collected from Landa Park in New Braunfels, TX. The temperatures was 24Â°C, and the soil was wet. The sample was taken 2 cm under surface.',\n",
       " 'isolation_temperature': 24,\n",
       " 'growth_notes': '',\n",
       " 'pcluster': {'cluster': 'BL', 'temperate': True},\n",
       " 'psubcluster': None,\n",
       " 'morphotype': 'SIPHO',\n",
       " 'plaque_notes': 'the diameter is 0.5 mm. It makes cloudy zone in the spot test.',\n",
       " 'lysogeny_notes': '',\n",
       " 'seq_finished': True,\n",
       " 'date_finished': '2017-03-06',\n",
       " 'seq_facility': {'short_name': 'UNT',\n",
       "  'facility_name': 'University of North Texas',\n",
       "  'contact_person': 'Richard Hale/Lee Hughes',\n",
       "  'shipping_address': '',\n",
       "  'facility_website': ''},\n",
       " 'seq_method': {'short_name': 'ILLUMINA',\n",
       "  'method_name': 'Illumina Sequencing'},\n",
       " 'coverage': None,\n",
       " 'finishing_reactions': None,\n",
       " 'genome_length': 42890,\n",
       " 'gcpercent': '61.2',\n",
       " 'end_type': '',\n",
       " 'repeat_length': None,\n",
       " 'end_notes': '',\n",
       " 'overhang_length': 9,\n",
       " 'overhang_sequence': 'CGGCAGGGA',\n",
       " 'sequencing_notes': \"Because this phage was sequenced using a tagmentation library-prep protocol, we can't be 100% sure about the ends and overhang, but there is strong evidence that we have called the correct base 1 and overhang, partly through similarity to related phages.\",\n",
       " 'duplicate_notes': '',\n",
       " 'is_annotated': True,\n",
       " 'is_phamerated': True,\n",
       " 'num_ORFs': 61,\n",
       " 'num_tRNAs': 7,\n",
       " 'num_tmRNAs': None,\n",
       " 'submitted_dnam_file': None,\n",
       " 'submitted_minimal_dnam_file': 'https://phagesdb.org/media/annotation_submission/DNAMasterFiles/Rowa_Minimal_Final.dnam5',\n",
       " 'submitted_author_list': None,\n",
       " 'submitted_cover_sheet': None,\n",
       " 'submitted_annotation_date': '2017-11-25T00:30:31-05:00',\n",
       " 'qced_genbank_file': 'https://phagesdb.org/media/annotation_submission/GenBankFiles/Rowa_flat.txt',\n",
       " 'qced_genbank_file_date': '2017-11-25T00:30:31-05:00',\n",
       " 'isolation_host': {'genus': 'Streptomyces',\n",
       "  'species': 'azureus',\n",
       "  'strain_name': 'NRRL B-2655',\n",
       "  'strain_species': {'id': 23,\n",
       "   'genus': 'Streptomyces',\n",
       "   'species_name': 'azureus',\n",
       "   'species_genus': {'id': 4,\n",
       "    'genus_domain': 'Bacteria',\n",
       "    'genus_name': 'Streptomyces',\n",
       "    'genus_wikipedia': '',\n",
       "    'is_actinomycetales': True},\n",
       "   'species_wikipedia': ''},\n",
       "  'ATCC_ID': '14921',\n",
       "  'genbank_accession': '',\n",
       "  'strain_information': ''},\n",
       " 'verified_hosts': [],\n",
       " 'verified_non_hosts': [],\n",
       " 'host_species': {'id': 23,\n",
       "  'genus': 'Streptomyces',\n",
       "  'species_name': 'azureus',\n",
       "  'species_genus': {'id': 4,\n",
       "   'genus_domain': 'Bacteria',\n",
       "   'genus_name': 'Streptomyces',\n",
       "   'genus_wikipedia': '',\n",
       "   'is_actinomycetales': True},\n",
       "  'species_wikipedia': ''},\n",
       " 'host_genus': {'id': 4,\n",
       "  'genus_domain': 'Bacteria',\n",
       "  'genus_name': 'Streptomyces',\n",
       "  'genus_wikipedia': '',\n",
       "  'is_actinomycetales': True},\n",
       " 'in_paper': False,\n",
       " 'paper_published_in': None,\n",
       " 'in_genbank': True,\n",
       " 'genbank_accession': 'MG593803',\n",
       " 'refseq_accession': '',\n",
       " 'plaque_file': 'https://phagesdb.org/media/plaquePics/Rowa_Plaque.jpg',\n",
       " 'em_file': 'https://phagesdb.org/media/emPics/Rowa_EMPic_h1BrzIx.jpg',\n",
       " 'digest_file': 'https://phagesdb.org/media/digestPics/Rowa_Digest.jpg',\n",
       " 'vdigest_file': None,\n",
       " 'fasta_file': 'https://phagesdb.org/media/fastas/Rowa.fasta',\n",
       " 'plaque_thumb_file': 'https://phagesdb.org/media/plaquePics/Rowa_Plaque_Thumb.png',\n",
       " 'em_thumb_file': 'https://phagesdb.org/media/emPics/Rowa_EMPic_Thumb.jpg',\n",
       " 'dnam_file': None,\n",
       " 'date_added': '2016-05-25T07:11:42-04:00',\n",
       " 'date_modified': '2021-11-26T05:43:11-05:00'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_phages[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3b58634f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "new_data = []\n",
    "\n",
    "for x in valid_phages:\n",
    "    row = []\n",
    "    row.append(x['phage_name'])\n",
    "    row.append(x[\"genome_length\"])\n",
    "    row.append(x[\"em_file\"])\n",
    "    row.append(x[\"p_institution\"]['institution_code'])\n",
    "    row.append(x[\"isolation_host\"][\"species\"])\n",
    "    new_data.append(row)\n",
    "    \n",
    "pd.DataFrame(new_data, columns = [\"Phage Name\", \"Genome Size\", \"TEM Image\", \"School\", \"Species\"]).to_csv(\"phage_images.csv\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cbc7a65e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UNTD\n",
      "UNTD\n",
      "WUSL\n",
      "WUSL\n",
      "WUSL\n",
      "UMBC\n",
      "UNTD\n",
      "UNTD\n",
      "LEHI\n",
      "UEVN\n",
      "UNTD\n",
      "UEVN\n",
      "UNTD\n",
      "UNTD\n",
      "UEVN\n",
      "UNTD\n",
      "UEVN\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n",
      "UMBC\n",
      "UMBC\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n",
      "LEHI\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n",
      "LEHI\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n",
      "WILK\n",
      "WUSL\n",
      "WUSL\n",
      "WUSL\n",
      "WUSL\n",
      "WUSL\n",
      "WUSL\n",
      "WUSL\n",
      "WUSL\n",
      "PITT\n",
      "WUSL\n",
      "WILK\n",
      "WUSL\n",
      "WUSL\n",
      "WUSL\n",
      "WUSL\n",
      "WUSL\n",
      "UMBC\n",
      "UMBC\n",
      "UMBC\n",
      "UMBC\n",
      "UMBC\n",
      "UMBC\n",
      "UMBC\n",
      "UMBC\n",
      "UMBC\n",
      "UMBC\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n",
      "UMBC\n",
      "UMBC\n",
      "UMBC\n",
      "UMBC\n",
      "UMBC\n",
      "UMBC\n",
      "UMBC\n",
      "UMBC\n",
      "UMBC\n",
      "UMBC\n",
      "UMBC\n",
      "UMBC\n",
      "UNTD\n",
      "WUSL\n",
      "WUSL\n",
      "WUSL\n",
      "WUSL\n",
      "WUSL\n",
      "WUSL\n",
      "WUSL\n",
      "WUSL\n",
      "WUSL\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n",
      "UNTD\n"
     ]
    }
   ],
   "source": [
    "for x in valid_phages:\n",
    "    print(x[\"p_institution\"]['institution_code'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77dbf8d4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "706a5b98",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd3ab683",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb9305fd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d469a15f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
